### PR TITLE
Added packages for ABB IRB 120 robotic manipulator

### PR DESCRIPTION
**Pull request by culletom**
As commented, this request is to add the abb_irb120_support and abb_irb120_moveit_config packages to the abb_experimental repository. Packages have been developed for ROS Hydro, and are structured with reference to the templates online.

### CODE REVIEW

File: abb_irb120_moveit_config/package.xml [hunk 1]
```diff
@@ -0,0 +1,41 @@
+<package>
+  <name>abb_irb120_moveit_config</name>
+  <version>0.2.0</version>
+  <description>
+    <p>
+      MoveIt package for the ABB IRB 120.
+    </p>
+    <p>
+      An automatically generated package with all the configuration and launch
+      files for using the ABB IRB 2400 with the MoveIt Motion Planning
```
Comment (shaun-edwards) [inline abb_irb120_moveit_config/package.xml]: Reference to IRB 2400 (copy and paste error?)

**Comment by shaun-edwards** [comment]
There appears to be variants of the IRB 120 (see [here](http://new.abb.com/products/robotics/industrial-robots/irb-120/irb-120-data)).  What variant does the PR cover.  Also, see issue ros-industrial/abb#47 for a description of variants.  See the fanuc [LRMate200ic](https://github.com/ros-industrial/fanuc/tree/hydro-devel/fanuc_lrmate200ic_support) for an example of a package with variants.

**Comment by shaun-edwards** [comment]
@Levi-Armstrong, please review as well.

File: abb_irb120_moveit_config/package.xml (hunk 1 above)
Comment (culletom) [inline abb_irb120_moveit_config/package.xml]: Yes just a copy and paste error, apologies

File: abb_irb120_moveit_config/config/joint_limits.yaml [hunk 2]
```diff
@@ -0,0 +1,31 @@
+joint_limits:
+  joint_1:
+    has_velocity_limits: true
+    max_velocity: 4.36332
+    has_acceleration_limits: false
+    max_acceleration: 0
```
Comment (gavanderhoorn) [inline abb_irb120_moveit_config/config/joint_limits.yaml]: Missing acceleration limits. I've seen strange planning results from MoveIt without them, but I don't know what the current state of this is.

I've always just either set them to the real limits (if available), or used the 'old way' of setting them to `1/5 * joint_vel_limit`.

File: abb_irb120_moveit_config/config/kinematics.yaml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+manipulator:
+  #kinematics_solver: abb_irb120_manipulator_kinematics/IKFastKinematicsPlugin
+  kinematics_solver: kdl_kinematics_plugin/KDLKinematicsPlugin
+  kinematics_solver_search_resolution: 0.005
+  kinematics_solver_timeout: 0.005
+  kinematics_solver_attempts: 3
```
Comment (gavanderhoorn) [inline abb_irb120_moveit_config/config/kinematics.yaml:5]: (optional) this moveit cfg has no ikfast plugin. We may want to log an issue for that.

File: abb_irb120_moveit_config/package.xml (hunk 1 above)
Comment (gavanderhoorn) [inline abb_irb120_moveit_config/package.xml]: Note also the 2400 references later on in this file.

File: abb_irb120_moveit_config/package.xml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    </p>
+  </description>
+  <author email="assistant@moveit.ros.org">MoveIt Setup Assistant</author>
+  <maintainer email="culletom@tcd.ie">Mark Culleton (Trinity College Dublin)</maintainer>
+
+  <license>BSD</license>
+
+  <url type="website">http://moveit.ros.org/</url>
+  <url type="bugtracker">https://github.com/ros-industrial/abb/issues</url>
+  <url type="repository">https://github.com/ros-industrial/abb</url>
```
Comment (gavanderhoorn) [inline abb_irb120_moveit_config/package.xml]: package is in `abb_experimental`, not in `abb`.

File: abb_irb120_moveit_config/package.xml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  <run_depend>moveit_ros_visualization</run_depend>
+  <run_depend>joint_state_publisher</run_depend>
+  <run_depend>robot_state_publisher</run_depend>
+  <run_depend>moveit_ros_move_group</run_depend>
+
+  <run_depend>xacro</run_depend>
+
+  <run_depend>industrial_robot_simulator</run_depend>
+  <run_depend>abb_irb2400_support</run_depend>
+  <run_depend>abb_irb2400_moveit_plugins</run_depend>
```
Comment (gavanderhoorn) [inline abb_irb120_moveit_config/package.xml]: without an ikfast plugin for the irb120, this dependency is unneeded (after resolving the 2400 copy/pasta).

File: abb_irb120_support/launch/load_irb120.launch [hunk 6]
```diff
@@ -0,0 +1,3 @@
+<launch>
+	<param name="robot_description" command="$(find xacro)/xacro.py '$(find abb_irb120_support)/urdf/irb120.xacro'" />
```
Comment (gavanderhoorn) [inline abb_irb120_support/launch/load_irb120.launch]: indent: use spaces (2 per level).

File: abb_irb120_support/package.xml [hunk 7]
```diff
@@ -0,0 +1,41 @@
+<package>
+  <name>abb_irb120_support</name>
+  <version>0.2.0</version>
+  <description>
+    <p>
+      ROS-Industrial support for the ABB IRB 120
```
Comment (gavanderhoorn) [inline abb_irb120_support/package.xml]: Missing dot at end of line.

File: abb_irb120_support/package.xml [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    </p>
+    <p>
+      This package contains configuration data, 3D models and launch files
+      for ABB IRB 120 manipulator. This currently includes the base model.
```
Comment (gavanderhoorn) [inline abb_irb120_support/package.xml]: Perhaps add 'only': "This currently includes the base model only.".

File: abb_irb120_support/package.xml [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    </p>
+    <p>
+      Joint limits and max joint velocities are based on the information in
+      the ABB data sheets.  All URDFs / XACROs are based on the
```
Comment (gavanderhoorn) [inline abb_irb120_support/package.xml]: Which data sheet in particular did you use? Ideally we'd provide a link (or document nr), so users can look it up and check their own robot.

File: abb_irb120_support/package.xml [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+      robot model and configuration you intend to use them with.
+    </p>
+  </description>
+  <author email="culletom@tcd.ie">Mark Culleton (Trinity College Dublin)</author>
+  <maintainer email="culletom@tcd.ie">Mark Culleton (Trinity College Dublin)</maintainer>
+  <license>Apache2</license>
+
+  <url type="website">http://moveit.ros.org/</url>
+  <url type="bugtracker">https://github.com/ros-industrial/abb/issues</url>
+  <url type="repository">https://github.com/ros-industrial/abb</url>
```
Comment (gavanderhoorn) [inline abb_irb120_support/package.xml]: again `abb` vs `abb_experimental`.

File: abb_irb120_support/package.xml [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+      Before using any of the configuration files and / or meshes included
+      in this package, be sure to check they are correct for the particular
+      robot model and configuration you intend to use them with.
+    </p>
+  </description>
+  <author email="culletom@tcd.ie">Mark Culleton (Trinity College Dublin)</author>
+  <maintainer email="culletom@tcd.ie">Mark Culleton (Trinity College Dublin)</maintainer>
+  <license>Apache2</license>
+
+  <url type="website">http://moveit.ros.org/</url>
```
Comment (gavanderhoorn) [inline abb_irb120_support/package.xml]: This is a support package, not a MoveIt package, right? `website` should probably be something like `http://wiki.ros.org/abb_irb120_support`.

File: abb_irb120_support/urdf/irb120.xacro [hunk 12]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" ?>
+
+<robot name="abb_irb120" xmlns:xacro="http://ros.org/wiki/xacro">
+	<xacro:include filename="$(find abb_irb120_support)/urdf/irb120_macro.xacro"/>
+	<xacro:abb_irb120 prefix=""/>
```
Comment (gavanderhoorn) [inline abb_irb120_support/urdf/irb120.xacro]: indent (see earlier comment).

File: abb_irb120_support/urdf/irb120_macro.xacro [hunk 13]
```diff
@@ -0,0 +1,195 @@
+<?xml version="1.0"?>
+<robot xmlns:xacro="http://ros.org/wiki/xacro">
+<xacro:macro name="abb_irb120" params="prefix">
+	<!-- link list -->
+	<link name="${prefix}base_link">
+		<visual>
+			<origin xyz="0 0 0" rpy="0 0 0"/>
+			<geometry>
+				<mesh filename="package://abb_irb120_support/meshes/irb120/visual/base_link.STL"/>
```
Comment (gavanderhoorn) [inline abb_irb120_support/urdf/irb120_macro.xacro]: I know the solidworks to urdf exporter does this by default (I don' t know if you've used it), but we try to keep all file extensions lower case.

File: abb_irb120_support/urdf/irb120_macro.xacro [hunk 14]
```diff
@@ -0,0 +1,195 @@
+<?xml version="1.0"?>
+<robot xmlns:xacro="http://ros.org/wiki/xacro">
+<xacro:macro name="abb_irb120" params="prefix">
+	<!-- link list -->
```
Comment (gavanderhoorn) [inline abb_irb120_support/urdf/irb120_macro.xacro]: Please change the indentation to two spaces per level. I know the other xacros don't do this (yet), but we should try to do this in the new ones. I've used [expand](http://linux.die.net/man/1/expand) for this earlier.

File: abb_irb120_support/urdf/irb120_macro.xacro [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+			<material name="black">
+			    <color rgba="0 0 0 1"/>
+			</material>	
+		</visual>
+		<collision>
+			<origin xyz="0 0 0" rpy="0 0 0"/>
+			<geometry>
+				<mesh filename="package://abb_irb120_support/meshes/irb120/collision/link_6.STL"/>
+			</geometry>
+            <material name="yellow"/>
```
Comment (gavanderhoorn) [inline abb_irb120_support/urdf/irb120_macro.xacro]: this seems like it's not correctly indented (also for other material elements earlier)?

File: abb_irb120_support/urdf/irb120_macro.xacro [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+
+	<!-- joint list -->
+	<joint name="${prefix}joint_1" type="revolute">
+		<origin rpy="0 0 0" xyz="0 0 0"/>
+		<parent link="${prefix}base_link"/>
+		<child link="${prefix}link_1"/>
+		<limit effort="0" lower="-2.87979" upper="2.87979" velocity="4.36332"/>
+		<axis xyz="0 0 1"/>
+        <calibration rising="0" falling="0"/>
+		<dynamics damping="0" friction="0"/>
```
Comment (gavanderhoorn) [inline abb_irb120_support/urdf/irb120_macro.xacro]: Not sure whether we want to keep these elements in a basic urdf. Do you have any specific use for them?

@shaun-edwards ?

**Comment by gavanderhoorn** [comment]
@shaun-edwards: this appears to model the 'base' IRB 120, which is also mentioned in the support package manifest (see [datasheet](http://www05.abb.com/global/scot/scot241.nsf/veritydisplay/3bd625bab3c7cae1c1257a0800495fac/$file/ROB0149EN_D_LR.pdf)). So I think the use of `abb_irb120` is correct, and does not need any further suffixes.

File: abb_irb120_moveit_config/config/joint_limits.yaml (hunk 2 above)
Comment (culletom) [inline abb_irb120_moveit_config/config/joint_limits.yaml]: I have contacted ABB about getting real acceleration limits, but for the minute I have used the 'old way' of setting them as above.

File: abb_irb120_support/urdf/irb120_macro.xacro (hunk 16 above)
Comment (culletom) [inline abb_irb120_support/urdf/irb120_macro.xacro]: I can remove these if needs be, I was just following the IRB 2400 description

**Comment by gavanderhoorn** [comment]
@culletom: it seems the PR now introduces meshes with both lowercase and uppercase extensions, in both `collision` and the `visual` subdirs?

**Comment by culletom** [comment]
Yes just to confirm that this is for the base IRB 120 model. As far as I can tell, the only difference between the two variants is their joint speed limits, so I will create the required files for the IRB 120T model as well, and will arrange with reference to the [fanuc LRMate200ic](https://github.com/ros-industrial/fanuc/tree/hydro-devel/fanuc_lrmate200ic_support) package

File: abb_irb120_moveit_config/package.xml [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+      Framework.
+    </p>
+  </description>
+  <author email="assistant@moveit.ros.org">MoveIt Setup Assistant</author>
+  <maintainer email="culletom@tcd.ie">Mark Culleton (Trinity College Dublin)</maintainer>
+
+  <license>BSD</license>
+
+  <url type="website">http://moveit.ros.org/</url>
+  <url type="bugtracker">https://github.com/ros-industrial/abb/issues</url>
```
Comment (gavanderhoorn) [inline abb_irb120_moveit_config/package.xml]: I'm not sure, but perhaps this should point to the issue tracker of the `abb_experimental` repository? All other experimental packages do so, I believe.

File: abb_irb120t_moveit_config/package.xml [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+      Framework.
+    </p>
+  </description>
+  <author email="assistant@moveit.ros.org">MoveIt Setup Assistant</author>
+  <maintainer email="culletom@tcd.ie">Mark Culleton (Trinity College Dublin)</maintainer>
+
+  <license>BSD</license>
+
+  <url type="website">http://moveit.ros.org/</url>
+  <url type="bugtracker">https://github.com/ros-industrial/abb/issues</url>
```
Comment (gavanderhoorn) [inline abb_irb120t_moveit_config/package.xml]: See earlier comment.

File: abb_irb120_support/package.xml [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+      in this package, be sure to check they are correct for the particular
+      robot model and configuration you intend to use them with.
+    </p>
+  </description>
+  <author email="culletom@tcd.ie">Mark Culleton (Trinity College Dublin)</author>
+  <maintainer email="culletom@tcd.ie">Mark Culleton (Trinity College Dublin)</maintainer>
+  <license>Apache2</license>
+
+  <url type="website">http://wiki.ros.org/abb</url>
+  <url type="bugtracker">https://github.com/ros-industrial/abb/issues</url>
```
Comment (gavanderhoorn) [inline abb_irb120_support/package.xml]: See earlier comment.

**Comment by gavanderhoorn** [comment]
@culletom: this looks good (btw: reformatting of the MoveIt pkg launch files was not necessary: they are autogenerated like that, and it is too much work to keep doing that every time something changes).

The `visual` meshes could probably be decimated a bit, but that is not important for now (we can log an issue).

I'd still rather remove the `dynamics` and `calibration` elements from the xacros (as they are never used), but let's hear some other opinions. @shaun-edwards, @Levi-Armstrong?

**Comment by Levi-Armstrong** [comment]
I agree, if they are never used it is probably best to remove the dynamics and calibration elements from xacros.

**Comment by gavanderhoorn** [comment]
Well, they can be used, but only in simulation contexts, which none of the 'basic' ros-i urdfs currently support / are setup for.

Personally I like to separate responsibilities as much as possible, so if the simulation properties (elements) could be added in some other way, that would have my preference. If that is not possible however, then we could leave them in.

**Comment by culletom** [comment]
Ok those elements are removed from the xacros, hopefully the packages are now taking shape. Apologies for so many mistakes, I'll make sure to keep the above in mind when creating ones in the future.  
As an aside, is there a reason why current ABB packages don't support simulation? Just asking as I'm currently working on Gazebo packages for the IRB 120, so could potentially upload something if it would be useful.

**Comment by gavanderhoorn** [comment]
@culletom: no need to apologise for anything. I can be a difficult person when it comes to these reviews, and it's not like we've published these conventions anywhere (or at least haven't covered all the details).

Thank _you_ for taking the time to contribute these packages.

> As an aside, is there a reason why current ABB packages don't support simulation? Just asking as I'm currently working on Gazebo packages for the IRB 120, so could potentially upload something if it would be useful. 

The only reason (if we can call it that) is that no one has really had a need for Gazebo compatibility until now, I think. So: yes, PRs adding Gazebo packages for the various robot models in the ABB support packages would be very much welcomed.

**Comment by Levi-Armstrong** [comment]
@gavanderhoorn: I believe all requested changes have been implemented, do you agree?

**Comment by gavanderhoorn** [comment]
I think so, yes. We can merge this and fix anything we missed in later PRs.

**Comment by culletom** [comment]
That is great to hear, I appreciate all the feedback. The only thing that has been brought up that hasn't been addressed is the decimation of the visual links, but as suggested this can be logged as an issue - I will try do this over the next few days, as well as cleaning up the gazebo packages that I am currently working on.

**Comment by gavanderhoorn** [comment]
Hm, seems I missed one additional thing. Not that important right now (as all ABB support pkgs have similar issues), but the translations between the joints in the urdfs seem 'odd'. While the final transform to `tool0` is probably correct, I wonder whether fi the frame for `joint_4` is at the right location (`<origin rpy="0 0 0" xyz="0.1496 -0.00475 0.07"`).

I'll log an issue for this.

**Comment by gavanderhoorn** [comment]
But really: thanks for contributing @culletom, very much appreciated.

**Pull request merged**